x-pack/plugins/canvas/public/components/function_reference_generator/function_reference_generator.tsx

```diff
@@ -0,0 +1,283 @@
+import { ArgumentType } from '../../../types';
+import { notifyService } from '../../services';
+import { isValidDataUrl, DATATABLE_COLUMN_TYPES } from '../../../common/lib';
+import { getFunctionExamples, FunctionExample } from './examples';
+
+const ALPHABET = 'abcdefghijklmnopqrstuvwxyz'.split('');
+const REQUIRED_ARG_ANNOTATION = '***';
+const MULTI_ARG_ANNOTATION = '†';
+const UNNAMED_ARG = '_Unnamed_';
+const ANY_TYPE = '`any`';
+
+const examplesDict = getFunctionExamples();
+
+const fnList = [
+  ...browserFunctions.map((fn) => fn().name),
+  ...serverFunctions.map((fn) => fn().name),
+  'asset',
+  'filters',
+  'timelion',
+  'to',
+  'font',
+  'var',
+  'var_set',
+  // ignore unsupported embeddables functions for now
+].filter((fn) => !['savedSearch'].includes(fn));
+
+interface FunctionDictionary {
+  [key: string]: AnyExpressionFunctionDefinition[];
+}
+
+const wrapInBackTicks = (str: string) => `\`${str}\``;
+const wrapInDoubleQuotes = (str: string) => (str.contains('"') ? str : `"${str}"`);
+const stringSorter = (a: string, b: string) => {
+  if (a < b) {
+    return -1;
+  }
+  if (a > b) {
+    return 1;
+  }
+  return 0;
+};
+
+const addFunctionLinks = (help: string, options?: { ignoreList?: string[] }) => {
+  const { ignoreList = [] } = options || {};
+  fnList.forEach((name: string) => {
+    const nameWithBackTicks = wrapInBackTicks(name);
+
+    // ignore functions with the same name as data types, i.e. string, date
+    if (
+      !ignoreList.includes(name) &&
+      !DATATABLE_COLUMN_TYPES.includes(name) &&
+      help.includes(nameWithBackTicks)
+    ) {
+      help = help.replace(nameWithBackTicks, `<<${name}_fn>>`);
+    }
+  });
+
+  return help;
+};
+
+const createDocs = (functionDefinitions: AnyExpressionFunctionDefinition[]) => {
+  const functionDefs = functionDefinitions.filter((fn: AnyExpressionFunctionDefinition) =>
+    fnList.includes(fn.name)
+  );
+
+  const functionDictionary: FunctionDictionary = {};
+  functionDefs.forEach((fn: AnyExpressionFunctionDefinition) => {
+    const firstLetter = fn.name[0];
+
+    if (!functionDictionary[firstLetter]) {
+      functionDictionary[firstLetter] = [];
+    }
+
+    functionDictionary[firstLetter].push(fn);
+  });
+  return `[role="xpack"]
+[[canvas-function-reference]]
+== Canvas function reference
+
+Behind the scenes, Canvas is driven by a powerful expression language,
+with dozens of functions and other capabilities, including table transforms,
+type casting, and sub-expressions.
+
+The Canvas expression language also supports <<canvas-tinymath-functions>>, which
+perform complex math calculations.
+
+A ${REQUIRED_ARG_ANNOTATION} denotes a required argument.
+
+A ${MULTI_ARG_ANNOTATION} denotes an argument can be passed multiple times.
+
+${createAlphabetLinks(functionDictionary)}
+
+${createFunctionDocs(functionDictionary)}`;
+};
+
+const createAlphabetLinks = (functionDictionary: FunctionDictionary) => {
```

Comment:
nit: might make sense to put all these helper funcs into a lib or something somewhere. Feels a bit big for a component file

Reply:
I extract all the doc related functions into a separate file in the same component folder.